controllers/rabbitmqcluster_controller.go

```diff
@@ -419,16 +427,32 @@ func (r *RabbitmqClusterReconciler) runSetPluginsCommand(ctx context.Context, rm
 	return nil
 }
 
-// Adds an arbitrary annotation (rabbitmq.com/lastRestartAt) to the StatefulSet PodTemplate to trigger a StatefulSet restart
-// if builder requires StatefulSet to be updated.
-func (r *RabbitmqClusterReconciler) restartStatefulSetIfNeeded(
-	ctx context.Context,
-	builder resource.ResourceBuilder,
-	operationResult controllerutil.OperationResult,
-	rmq *rabbitmqv1beta1.RabbitmqCluster) (restarted bool) {
+// Adds an arbitrary annotation to the sts PodTemplate to trigger a sts restart
+// it compares annotation "rabbitmq.com/serverConfUpdatedAt" from server-conf configMap and annotation "rabbitmq.com/lastRestartAt" from sts
+// to determine whether to restart sts
+func (r *RabbitmqClusterReconciler) restartStatefulSetIfNeeded(ctx context.Context, rmq *rabbitmqv1beta1.RabbitmqCluster) (time.Duration, error) {
+	serverConf, err := r.configMap(ctx, rmq, rmq.ChildResourceName(resource.ServerConfigMap))
+	if err != nil {
+		// requeue request after 10s if unable to find server-conf configmap, else return the error
+		return 10 * time.Second, client.IgnoreNotFound(err)
+	}
```

Comment:
I'm thinking that we could rewrite this block with `retry.OnError()` [function](https://godoc.org/k8s.io/client-go/util/retry#OnError). It would require to rewrite the helper function so that it only returns an error, and gets the config map into a pointer argument. The main benefit I observe is avoiding a complete requeue of the reconcile function in the scenario where the ConfigMap is taking a bit of time to come up in the Informer, which I think it's not necessary in this case. In the event of an error, we can return the error in the result and the reconcile will requeue automatically.

Reply:
I see your point Aitor. I think a requeue is a rather cheap thing here compared to adding retry logic. Sure we will jump back to the beginning of the reconcile loop, however it's the same request that we are requeuing, and the request was already processed before. I think adding retry logic is more overhead in this case considering that objects records should be created instantly, so this Get is very unlikely to fail anyways. Thoughts?

Reply:
I guess it depends how we define 'cheap'. If we assume every function/check has cost 1 and running the complete `Reconcile()` function in its happy path (no errors) has cost N, then a requeue here will imply a cost of aproximately 2N, or 2x(3/4 x N) since we are almost at the end of the function. If we add retry logic and every retry is, guesstimate cost of 5, and in the worst case, the retry succeeds in the last attempt e.g. 4th, then the cost of the function would be N + 20, which is equivalent to N, or N + M; in this case, N + M is better than 2N if M < N, which I think it's the case.

Reply:
> I believe that going through the "update Status" and "resource builders" part is more expensive than retrying a Get() a number of times.

I agree that requeuing is going to be more expensive. The question here is what's the odds of the Get request to fail in this scenario? I would say that Get is **very unlikely** to fail with the error Not Found. The controller can successfully Get/List resources from the k8s api after creating them is an assumption in our controller code. If we decide that the api is unreliable to return resources, then we should have retries in all the Get request that we are performing. I don't see why we should optimize for a very unlikely scenario here.

Reply:
I don't feel strong about it. I simply think that if we can recover from an error by retrying instead of requeuing, it makes sense to do so. In any case, the outcome will be the same so 🤷‍♂️